### PR TITLE
support multi-dim reward for AC and PPO

### PR DESCRIPTION
**Pull request by hnyu**
For both, the multi-dim advantages are weighted averaged before pg_loss is computed. The td_loss is still multi-dim. 

Some other minor changes:
1. Added a Lagrangian method using prediction rewards.
2. Added an option of projecting the positive weights to a simplex before returning them.
3. Added a common ppo_conf.py file

### CODE REVIEW

File: alf/algorithms/actor_critic_loss.py [hunk 1]
```diff
@@ -130,10 +153,15 @@ def forward(self, info):
             advantages = torch.clamp(advantages, -self._advantage_clip,
                                      self._advantage_clip)
 
+        if info.reward_weights is not ():
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_loss.py]: It seems that now pytorch can correctly handle `tensor != ()`. So we don't need to use `is not`. `is not` will generate warning with python 3.9.

File: alf/algorithms/ppo_algorithm.py [hunk 2]
```diff
@@ -52,23 +52,30 @@ def train_step(self, inputs: TimeStep, state, rollout_info):
...
+            discounts=discounts,
             td_lambda=self._loss._lambda,
             time_major=False)
-        advantages = torch.cat([
-            advantages,
-            torch.zeros(*advantages.shape[:-1], 1, dtype=advantages.dtype)
-        ],
-                               dim=-1)
+        advantages = tensor_utils.tensor_extend_zero(
+            advantages.transpose(0, 1)).transpose(0, 1)
```
Comment (emailweixu) [inline alf/algorithms/ppo_algorithm.py]: If you want to change to use tensor_extend_zero, I feel it's better to enchance tensor_extend_zero to add a dim argument.

File: alf/algorithms/lagrangian_reward_weight_algorithm.py [hunk 3]
```diff
@@ -59,6 +69,9 @@ def __init__(self,
                 its init value is always used.
             optimizer (optimizer): optimizer for learning the reward weights.
             init_weights (float|list[float]): the initial reward weights.
+            max_weight (float): the reward weights will be clipped up to this value
+            reward_weight_projection (bool): whether project the weights to
```
Comment (emailweixu) [inline alf/algorithms/lagrangian_reward_weight_algorithm.py]: reward_weight_normalization might be a more descriptive name than reward_weight_projection

File: alf/algorithms/lagrangian_reward_weight_algorithm.py [hunk 4]
```diff
@@ -47,6 +54,9 @@ def __init__(self,
                  reward_thresholds,
                  optimizer,
                  init_weights=1.,
+                 max_weight=None,
+                 reward_weight_projection=True,
+                 lambda_transform=F.softplus,
```
Comment (emailweixu) [inline alf/algorithms/lagrangian_reward_weight_algorithm.py:61]: comment for lambda_transform. Only F.softplus and torch.exp are supported.

File: alf/algorithms/lagrangian_reward_weight_algorithm.py [hunk 5]
```diff
@@ -59,6 +69,9 @@ def __init__(self,
                 its init value is always used.
```
Comment (emailweixu) [inline alf/algorithms/lagrangian_reward_weight_algorithm.py]: This not exactly true when reward_weight_projection is used.

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/utils/tensor_utils.py [hunk 6]
```diff
@@ -75,16 +75,22 @@ def tensor_extend(x, y):
...
     Returns:
-        Tensor: the extended tensor. Its shape is ``(x.shape[0]+1, x.shape[1:])``
+        Tensor: the extended tensor. Its shape is
+            ``(*x.shape[:dim], x.shape[dim]+1, *x.shape[dim+1:])``
     """
-    return torch.cat((x, torch.zeros(1, *x.shape[1:], dtype=x.dtype)))
+    if dim < 0:
+        return tensor_extend_zero(x, dim + x.ndim)
+
+    zero_shape = x.shape[:dim] + (1, ) + x.shape[dim + 1:]
```
Comment (emailweixu) [inline alf/utils/tensor_utils.py]: Code can be simplified a little bit (line 88-89 can be removed):
```python
zero_shape = list(x.shape)
zero_shape[dim] = 1
```

File: alf/algorithms/lagrangian_reward_weight_algorithm.py [hunk 7]
```diff
@@ -79,52 +95,167 @@ def __init__(self,
...
+        """
+        Args:
+            reward_spec (TensorSpec): a rank-1 tensor spec representing multi-dim
+                rewards.
+            reward_thresholds (list[float|None]): a list of floating numbers,
+                each representing a desired minimum reward threshold in expectation.
+                If any entry is None, then that reward weight won't be tuned and
+                its init value (after normalization if ``reward_weight_normalization=True``)
+                is always used. If a single floating value, it's shared by all
+                weights.
```
Comment (emailweixu) [inline alf/algorithms/lagrangian_reward_weight_algorithm.py]: The comment for base class also needs to be updated.

**Review by emailweixu — APPROVED** [review]
